source/core/module_coreHelpers.js

```diff
@@ -75,33 +104,68 @@
                     str += String.fromCharCode(u0);
                     continue;
                 }
+
+                // Look ahead to validate the UTF-8 structure
+                if ((u0 & 0xE0) === 0xC0) {
```

Comment:
Yea talked about it with craig, memo, and jesse and agree that for now will add additional comments to help clarify the behavior of each region and expected byte structures.

I will also create a separate issue for breaking up the code into more components including benchmarking. Right now I'm worried that sub function calls may result in a performance penalty so we should make that change when we can benchmark for regressions as well. We should also investigate usage of the [JavaScript TextDecoder API](https://www.w3.org/TR/encoding/#dom-textdecoder) which may be faster but might only support valid utf-8 sequences and may have varying crossbrowser behavior